File: frontend/gamer_circle/lib/core/amenities.dart
/// Amenity bitmask — keep in sync with backend `app/core/amenities.py`.
class Amenity {
  static const int ps5 = 1;
  static const int pc = 2;
  static const int vr = 4;
  static const int snooker = 8;
  static const int ac = 16;
  static const int parking = 32;
  static const int cafe = 64;

  static const Map<int, String> labels = {
    ps5: 'PS5',
    pc: 'PC',
    vr: 'VR',
    snooker: 'Snooker',
    ac: 'AC',
    parking: 'Parking',
    cafe: 'Cafe',
  };

  static List<String> namesFromMask(int mask) =>
      labels.entries.where((e) => mask & e.key != 0).map((e) => e.value).toList();

  static int maskFromNames(Iterable<String> names) {
    final lookup = {
      for (final e in labels.entries) e.value.toLowerCase(): e.key,
      'playstation': ps5,
      'pool': snooker,
      'billiards': snooker,
    };
    var mask = 0;
    for (final n in names) {
      final bit = lookup[n.trim().toLowerCase()];
      if (bit != null) mask |= bit;
    }
    return mask;
  }
}

File: backend/app/core/amenities.py
"""Amenity bitmask for discovery filters.

Filter predicate: ``amenities_mask & :mask = :mask`` (no join).
Keep in sync with ``frontend/gamer_circle/lib/core/amenities.dart``.
"""

from __future__ import annotations

from enum import IntFlag


class Amenity(IntFlag):
    PS5 = 1
    PC = 2
    VR = 4
    SNOOKER = 8
    AC = 16
    PARKING = 32
    CAFE = 64


AMENITY_LABELS: dict[Amenity, str] = {
    Amenity.PS5: "PS5",
    Amenity.PC: "PC",
    Amenity.VR: "VR",
    Amenity.SNOOKER: "Snooker",
    Amenity.AC: "AC",
    Amenity.PARKING: "Parking",
    Amenity.CAFE: "Cafe",
}


def mask_from_names(names: list[str] | None) -> int:
    if not names:
        return 0
    lookup = {v.lower(): int(k) for k, v in AMENITY_LABELS.items()}
    # aliases
    lookup.update(
        {
            "playstation": int(Amenity.PS5),
            "playstation 5": int(Amenity.PS5),
            "pool": int(Amenity.SNOOKER),
            "billiards": int(Amenity.SNOOKER),
        }
    )
    mask = 0
    for name in names:
        bit = lookup.get(name.strip().lower())
        if bit:
            mask |= bit
    return mask


def names_from_mask(mask: int) -> list[str]:
    return [label for flag, label in AMENITY_LABELS.items() if mask & int(flag)]
